src/Particular.ThroughputCollector/AuditThroughput/AuditThroughputCollectorHostedService.cs

```diff
@@ -1,104 +1,120 @@
+        backgroundTimer = timeProvider.CreateTimer(async _ => await GatherThroughput(stoppingToken), null, TimeSpan.FromSeconds(30), TimeSpan.FromDays(1));
+        stoppingToken.Register(_ => backgroundTimer?.Dispose(), null);
 
-            return Task.CompletedTask;
-        }
+        return Task.CompletedTask;
+    }
+
+    async Task GatherThroughput(CancellationToken cancellationToken)
+    {
+        var utcYesterday = DateOnly.FromDateTime(timeProvider.GetUtcNow().DateTime).AddDays(-1);
+        logger.LogInformation($"Gathering throughput from audit for {utcYesterday.ToShortDateString}");
 
-        async Task GatherThroughput()
+        try
         {
-            var utcYesterday = DateOnly.FromDateTime(timeProvider.GetUtcNow().DateTime).AddDays(-1);
-            logger.LogInformation($"Gathering throughput from audit for {utcYesterday.ToShortDateString}");
+            await VerifyAuditInstances(cancellationToken);
+
+            var knownEndpoints = await AuditCommands.GetKnownEndpoints(endpointsApi);
+            var knownEndpointsLookup = knownEndpoints
+                .ToDictionary(knownEndpoint => new EndpointIdentifier(knownEndpoint.Name, ThroughputSource.Audit));
 
-            try
+            if (!knownEndpoints.Any())
             {
-                await VerifyAuditInstances();
+                logger.LogWarning("No known endpoints could be found.");
+            }
 
-                var knownEndpoints = await AuditCommands.GetKnownEndpoints(endpointsApi);
+            foreach (var tuple in await dataStore.GetEndpoints(knownEndpointsLookup.Keys, cancellationToken))
+            {
+                var endpointId = tuple.Id;
+                var endpoint = tuple.Endpoint;
 
-                if (!knownEndpoints.Any())
+                if (endpoint?.LastCollectedDate == utcYesterday)
                 {
-                    logger.LogWarning("Successfully connected to ServiceControl API but no known endpoints could be found.");
+                    continue;
                 }
 
-                foreach (var endpoint in knownEndpoints)
+                var auditCounts = (await AuditCommands.GetAuditCountForEndpoint(auditCountApi, knownEndpointsLookup[endpointId].UrlName)).ToList();
+
+                if (endpoint == null)
                 {
-                    if (!await ThroughputRecordedForYesterday(endpoint.Name, utcYesterday))
-                    {
-                        var auditCounts = await AuditCommands.GetAuditCountForEndpoint(auditCountApi, endpoint.UrlName);
-                        //for each endpoint record the audit count for the day we are currently doing as well as any others that are available
-                        await dataStore.RecordEndpointThroughput(SCEndpointToEndpoint(endpoint, auditCounts));
-                    }
+                    endpoint = ConvertToEndpoint(knownEndpointsLookup[endpointId], auditCounts);
+                    await dataStore.SaveEndpoint(endpoint, cancellationToken);
```

Comment:
Why do we need to do this Save here, if the RecordEndpointThroughput method which is called below will add the endpoint if it doesn't exist? I think we should only have one of the methods on the datastore - it's unnecessarily confusing having the two.

Reply:
My plan is to ultimately remove the side effect of creating the endpoint in RecordEndpointThroughput.  The endpoint has other parameters (e.g. sanitized name, indicators, etc) which do not make sense to pass into RecordEndpointThroughput.

Reply:
AppendEndpointThroughput also does the adding of an endpoint if it doesn't exist. I would like to clean all that up so that it's very clear as to what is happening in every method, without making too many db calls.
I'd be happy with just the 2 methods:

- CreateEndpointThroughput (which only creates it ensuring that it doesn't overwrite any existing ones)
- AppendEndpointThroughput (which appends or creates it if it doesn't exist)

I don't have an issue with both methods also adding the endpoint itself if it doesn't exist - it is part of the one record in the db afterall.